Replace `FreeTypeFaceHelpers` with a safe wrapper struct

simonwuelker opened the pull request:
Testing: only safety annotations change, no tests are required.
Fixes: https://github.com/servo/servo/issues/38627

simonwuelker committed:
Mark FreeTypeFaceHelper methods as unsafe

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>

mrobinson commented:
> I'm not sure what you mean by "guard the face with FT_Reference_Face", but a `NonNull<FT_Face>` can still be dangling, aliased and unaligned. Maybe we should consider using a safe wrapper library.

If we get the pointer directly from Freetype and ensure the face itself isn't freed (by using RAII to call `FT_Reference_Face` and `FT_Done_Face`, isn't the only really unsafe operation where we place the pointer into a `NonNull` wrapper? In what cases can Freetype give us an unaligned pointer?

simonwuelker commented:
> > If we get the pointer directly from Freetype and ensure the face itself isn't freed (by using RAII to call `FT_Reference_Face` and `FT_Done_Face`), isn't the only really unsafe operation where we place the pointer into a `NonNull` wrapper? In what cases can Freetype give us an unaligned pointer?
> 
> Freetype won't give us unaligned pointers (I hope), but the function signatures right now don't guarantee that the pointer actually came from freetype. I can write simple wrapper that does what you described 👍 .

Done, the wrapper is `FreetypeFace` and ensures the `FT_Face` pointer is nonnull and came from `FT_New_Face`. All the methods from `FreeTypeHelpers` become associated functions on the wrapper.

mrobinson reviewed (approved):
Thanks! This looks like a nice wrapper.

simonwuelker commented:
We're running into a deadlock because the wrapper reliably frees the `FT_Face`, which previously wasn't happening (:

simonwuelker committed:
Put Freetype library singleton in a reentrant mutex

Destructors have to be able to lock the singleton
and there is no reasonable way to ensure that no lock
is held when destructors are called.

Signed-off-by: Simon Wülker <simon.wuelker@arcor.de>
